### PR TITLE
Refactoring `interfaces` in preparation for Jacobian Product work

### PR DESCRIPTION
**Pull request by albi3ro**
This PR accomplishes some preparation work for the change prototyped in #4368 .

The main changes are:

**Separate legacy and current ml boundaries from each other**

For example, `qml.interfaces.autograd.execute` will no longer call to `qml.interfaces.autograd._execute_legacy`.  `_execute_legacy` must manually be called by `qml.interfaces.execution._execute_legacy`.  This change is mandatory to change the call signature for `qml.interfaces.autograd.execute` without having to port the changes to the legacy system.

**Pull importing the ml boundary into a private helper function**

Under the new system, the ml boundary will no longer be responsible for figuring out whether or not the Jacobian products are differentiable.  Instead, the `gradient_fn`/ `vjp_fn` they are provided may or may not be differentiable, and the ml boundary does not need to care.

Having this helper function will make this unspooling of recursion easier.

**Make the inner `execute_fn` responsible for converting to numpy:**

Now, the ml boundaries are no longer responsible for converting the tapes to numpy.  This is mandatory since the frameworks will no longer know what happens inside `execute_fn` and whether or not it needs to be differentiable.

**Create `_make_inner_execute` helper function**

Non-mandatory. Just cleans things up a little bit.

**Identify jax-jit early:**

Currently, checks for jax jit are called in two places.  Now, we check for jax-jit early on, and then just check whether `interface=="jax-jit"`. Non-mandatory, but makes things a little eaiser.

### CODE REVIEW

File: pennylane/devices/qubit/sampling.py [hunk 1]
```diff
@@ -216,6 +216,7 @@ def sample_state(
         # rng.choice doesn't support broadcasting
         samples = np.stack([rng.choice(basis_states, shots, p=p) for p in probs])
     else:
+        print(probs)
```
Comment (albi3ro) [inline pennylane/devices/qubit/sampling.py:219]:
```suggestion
```

File: pennylane/optimize/qnspsa.py [hunk 2]
```diff
@@ -408,6 +408,7 @@ def _get_overlap_tape(self, cost, args1, args2, kwargs):
         op_forward = self._get_operations(cost, args1, kwargs)
         op_inv = self._get_operations(cost, args2, kwargs)
 
+        print(op_inv)
```
Comment (albi3ro) [inline pennylane/optimize/qnspsa.py:411]:
```suggestion
```

File: tests/transforms/test_batch_transform.py [hunk 3]
```diff
@@ -352,6 +352,7 @@ def circuit(x):
         spy = mocker.spy(circuit.device, "batch_execute")
         circuit(x)
         tapes = spy.call_args[0][0]
+        print(spy.call_args)
```
Comment (albi3ro) [inline tests/transforms/test_batch_transform.py:355]:
```suggestion
```

File: pennylane/interfaces/torch.py [hunk 4]
```diff
@@ -500,6 +486,7 @@ def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_d
 
 def _res_to_torch(r, ctx):
     """Convert results from unwrapped execution to torch."""
+    print(r)
```
Comment (albi3ro) [inline pennylane/interfaces/torch.py:489]:
```suggestion
```

File: pennylane/gradients/general_shift_rules.py [hunk 5]
```diff
@@ -396,8 +396,12 @@ def _copy_and_shift_params(tape, indices, shifts, multipliers, cast=False):
 
         # Shift copied parameter
         new_params = list(op.data)
-        new_params[p_idx] = new_params[p_idx] * qml.math.convert_like(multiplier, new_params[p_idx])
-        new_params[p_idx] = new_params[p_idx] + qml.math.convert_like(shift, new_params[p_idx])
+        multiplier = qml.math.convert_like(multiplier, new_params[p_idx])
```
Comment (albi3ro) [inline pennylane/gradients/general_shift_rules.py:399]: We are now passing interface-specific variables to the gradients module.  Without the `cast_like`, Tensorflow was raising errors.

File: pennylane/interfaces/execution.py [hunk 6]
```diff
@@ -454,8 +535,11 @@ def cost_fn(params, x):
         for tape in tapes:
             params.extend(tape.get_parameters(trainable_only=False))
         interface = qml.math.get_interface(*params)
+    if interface == "jax":
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:538]: Previously, we were running this check in two separate places.  Now we just do it once at the beginning.

File: pennylane/interfaces/execution.py [hunk 7]
```diff
@@ -474,56 +558,43 @@ def cost_fn(params, x):
...
-            cache,
-            expand_fn=expand_fn,
-            return_tuple=False,
-            pass_kwargs=new_device_interface,
-        )
-        results = cached_execute_fn(tapes, execution_config=config)
+        results = inner_execute(tapes)
         return batch_fn(results)
 
-    # the default execution function is batch_execute
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:514]: Moved into `_make_inner_execute`

File: pennylane/interfaces/execution.py [hunk 8]
```diff
@@ -88,6 +88,53 @@ def _adjoint_jacobian_expansion(
     return tapes
 
 
+def _get_ml_framework_boundary(interface, grad_on_execution):
+    """Imports and returns the function that binds derivatives of the required ml framework.
+
+    Args:
+        interface (str): The designated for the ml framework.
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:95]:
```suggestion
        interface (str): The designated ml framework.

```

File: pennylane/interfaces/execution.py [hunk 9]
```diff
@@ -474,56 +561,43 @@ def cost_fn(params, x):
...
-        batch_execute = set_shots(device, override_shots)(device.batch_execute)
-
     expand_fn = _preprocess_expand_fn(expand_fn, device, max_expansion)
 
+    # changing this set of conditions causes a bunch of tests to break.
+    no_interface_boundary_required = interface is None or gradient_fn in {None, "backprop"}
+    device_supports_interface_data = no_interface_boundary_required and (
+        interface is None
+        or gradient_fn == "backprop"
+        or getattr(device, "short_name", "") == "default.mixed"
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:571]: I guess `short_name` is not part of the new API. do we want it to be?

File: pennylane/interfaces/execution.py [hunk 10]
```diff
@@ -561,7 +641,10 @@ def gradient_fn(internal_tapes):
                     device: the device to execute on
                     config: the ExecutionConfig that specifies how to take the derivative.
                 """
-                return device.compute_derivatives(internal_tapes, config)
+                numpy_tapes = tuple(
+                    qml.transforms.convert_to_numpy_parameters(t) for t in internal_tapes
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:645]: starting to wonder if we should just make this accept a batch of tapes 💭

File: pennylane/interfaces/execution.py [hunk 11]
```diff
@@ -617,36 +714,15 @@ def gradient_fn(internal_tapes):
...
-
-        elif mapped_interface == "jax":
-            _execute = _get_jax_execute_fn(interface, tapes)
-
-        res = _execute(
+        res = ml_boundary(
             tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_diff=max_diff
         )
-
     except ImportError as e:
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:722]: doesn't the ImportError happen in `_get_ml_framework_boundary` (above) now?

File: pennylane/math/utils.py [hunk 12]
```diff
@@ -138,7 +138,9 @@ def cast_like(tensor1, tensor2):
     >>> cast_like(x, y)
     tensor([1., 2.])
     """
-    if not is_abstract(tensor2):
+    if isinstance(tensor2, ArrayBox):
+        dtype = ar.to_numpy(tensor2._value).dtype.type  # pylint: disable=protected-access
```
Comment (timmysilv) [inline pennylane/math/utils.py:142]: does this not break anything? why did this need to be added?

File: pennylane/interfaces/execution.py (hunk 11 above)
Comment (albi3ro) [reply pennylane/interfaces/execution.py:722]: I think the error is tested by mocking some functions in each library.

But the actual error would be raised on importing the relevant function.

File: pennylane/interfaces/jax.py [hunk 13]
```diff
@@ -458,7 +458,7 @@ def execute_wrapper_jvp(primals, tangents):
         """Primals[0] are parameters as Jax tracers and tangents[0] is a list of tangent vectors as Jax tracers."""
         if isinstance(gradient_fn, qml.gradients.gradient_transform):
             at_max_diff = _n == max_diff
-            new_tapes = set_parameters_on_copy_and_unwrap(tapes, primals[0], unwrap=at_max_diff)
+            new_tapes = set_parameters_on_copy_and_unwrap(tapes, primals[0], unwrap=False)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jax.py:461]: Now they are unwrapped inside the "inner execute function". 

But I didn't port the changes to legacy.

File: pennylane/math/utils.py (hunk 12 above)
Comment (albi3ro) [reply pennylane/math/utils.py:142]: It was breaking something, but now I'm forgetting what.

File: pennylane/interfaces/jax.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/interfaces/jax.py:461]: legacy... makes sense, ty!

**Comment by albi3ro** [comment]
[sc-43013]

File: pennylane/interfaces/execution.py (hunk 11 above)
Comment (timmysilv) [reply pennylane/interfaces/execution.py:722]: so can we remove this?

Comment (albi3ro) [reply pennylane/interfaces/execution.py:701]: Some tests would fail, and need to be removed. See: https://github.com/PennyLaneAI/pennylane/blob/6e8b491cab67f1c87de2a54b178c37653440069b/tests/interfaces/test_jax.py#L32

Comment (timmysilv) [reply pennylane/interfaces/execution.py:701]: is it too much to do in this PR?

File: doc/releases/changelog-dev.md [hunk 14]
```diff
@@ -107,6 +107,13 @@
 * The experimental `DefaultQubit2` device now supports computing VJPs and JVPs using the adjoint method.
   [(#4374)](https://github.com/PennyLaneAI/pennylane/pull/4374)
 
+* Refactoring of `pennylane/interfaces`.  The `execute_fn` passed to the machine learning framework boundaries 
+  is now responsible for converting parameters to numpy. The gradients module can now handle tensorflow parameters,
+  but gradient tapes now retain the original dtype instead of converting to float64.  This may cause instability
```
Comment (rmoyard) [inline doc/releases/changelog-dev.md:112]: Can you add the float64 part in breaking changes?

File: pennylane/interfaces/execution.py [hunk 15]
```diff
@@ -599,36 +693,15 @@ def gradient_fn(internal_tapes):
...
-                from .tensorflow import execute as _execute
-
-        elif mapped_interface == "torch":
-            from .torch import execute as _execute
-
-        elif mapped_interface == "jax":
-            _execute = _get_jax_execute_fn(interface, tapes)
-
-        res = _execute(
+        res = ml_boundary(
```
Comment (rmoyard) [inline pennylane/interfaces/execution.py:698]: I don't really agree with naming it ml_boundary, I think you should add the purpose of the function in the name maybe `ml_boundary_execute` or `ml_execute`. If you are not familiar with the codebase you cannot directly understand what happen to the tapes.

File: pennylane/interfaces/execution.py [hunk 16]
```diff
@@ -86,6 +86,54 @@ def _adjoint_jacobian_expansion(
     return tapes
 
 
+def _get_ml_framework_boundary(interface, grad_on_execution):
```
Comment (rmoyard) [inline pennylane/interfaces/execution.py:89]: Nice to extract it but same comment with naming as below.

File: tests/devices/qubit/test_sampling.py [hunk 17]
```diff
@@ -21,7 +21,7 @@
 from pennylane.devices.qubit import simulate
 from pennylane.devices.qubit import sample_state, measure_with_samples
 
-two_qubit_state = np.array([[0, 1j], [-1, 0]]) / np.sqrt(2)
+two_qubit_state = np.array([[0, 1j], [-1, 0]], dtype=np.complex128) / np.sqrt(2)
```
Comment (rmoyard) [inline tests/devices/qubit/test_sampling.py:24]: Is it not working without np.complex128?

File: tests/devices/test_default_mixed_tf.py [hunk 18]
```diff
@@ -617,7 +617,7 @@ def cost(params):
         phi = -0.234
         lam = 0.654
 
-        params = tf.Variable([theta, phi, lam], trainable=True)
+        params = tf.Variable([theta, phi, lam], trainable=True, dtype=tf.float64)
```
Comment (rmoyard) [inline tests/devices/test_default_mixed_tf.py:620]: For all the test where you added the precision? Was it failing, if yes we have to document the case and add it to the breaking changes. If no you can ignore this message.

**Review by rmoyard — COMMENTED** [review]
Thanks nice changes! :100: I have left some comments

File: pennylane/interfaces/execution.py [hunk 19]
```diff
@@ -163,6 +211,40 @@ def device_expansion_function(tape):  # pylint: disable=function-redefined
     return device_expansion_function
 
 
+def _make_inner_execute(
+    device, override_shots, cache, expand_fn=None, execution_config=None, numpy_only=True
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/execution.py:215]: For higher-order derivatives, the `execute_fn` passed to the ml boundary will be another ml boundary.  This function will only be the "inner most" just above the device.

File: pennylane/interfaces/execution.py (hunk 11 above)
Comment (albi3ro) [reply pennylane/interfaces/execution.py:705]: removing these import error tests. That's not where import errors are actually going to get raised.

File: doc/releases/changelog-dev.md (hunk 14 above)
Comment (albi3ro) [reply doc/releases/changelog-dev.md:160]: Adding. And adding a warning to `qml.gradients.finite_diff` if float32 is detected.

File: tests/devices/qubit/test_sampling.py (hunk 17 above)
Comment (albi3ro) [reply tests/devices/qubit/test_sampling.py:24]: complex64 is too inaccurate with finite diff.

File: tests/devices/test_default_mixed_tf.py (hunk 18 above)
Comment (albi3ro) [reply tests/devices/test_default_mixed_tf.py:620]: Added to breaking changes section and added warning to `finite_diff` if the data is float32.

File: pennylane/gradients/finite_difference.py [hunk 20]
```diff
@@ -336,6 +337,10 @@ def finite_diff(
             validate_params=validate_params,
             shots=shots,
         )
+
+    if any(qml.math.get_dtype_name(p) == "float32" for p in tape.get_parameters()):
+        warn("Finite differences with float32 detected. Answers may be inaccurate. ", UserWarning)
```
Comment (trbromley) [inline pennylane/gradients/finite_difference.py:342]: Nice! Please could we add a recommended action? E.g., it's not immediately clear what I'd need to do to stop the warning, such as to cast all my parameters to float64?

**Review by timmysilv — APPROVED** [review]
looks great! I'm also wondering about _when_ the 64-bit dtypes are required, because it seems (as it was for me while testing some things) that we just provide them when something failed without them.

File: pennylane/interfaces/execution.py [hunk 21]
```diff
@@ -438,8 +526,17 @@ def cost_fn(params, x):
         for tape in tapes:
             params.extend(tape.get_parameters(trainable_only=False))
         interface = qml.math.get_interface(*params)
+    if interface == "jax":
+        try:
+            from .jax import get_jax_interface_name
+        except ImportError as e:  # pragma: no-cover
+            raise qml.QuantumFunctionError(
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:533]:
```suggestion
            raise qml.QuantumFunctionError( # pragma: no-cover

```

**Review by rmoyard — APPROVED** [review]
It looks good to me :+1: I just want to be sure that @trbromley agrees with the finite diff/float changes before merging

**Comment by trbromley** [comment]
> It looks good to me +1 I just want to be sure that @trbromley agrees with the finite diff/float changes before merging

Thanks @rmoyard! This looks good to me, I think the warning nicely addresses the main concern around the finite-diff method. Would just like to resolve [this](https://github.com/PennyLaneAI/pennylane/pull/4415#discussion_r1290048575) comment, i.e., adding a bit more clarity for what users should do if they see this warning.

File: pennylane/gradients/finite_difference.py (hunk 20 above)
Comment (albi3ro) [reply pennylane/gradients/finite_difference.py]: added. now says to use float64.

File: pennylane/gradients/finite_difference.py [hunk 22]
```diff
@@ -329,6 +330,15 @@ def finite_diff(
...
             validate_params=validate_params,
         )
+
+    if any(qml.math.get_dtype_name(p) == "float32" for p in tape.get_parameters()):
+        warn(
+            "Finite differences with float32 detected. Answers may be inaccurate. float64 is recommended.",
+            UserWarning,
+        )
+
+    shots = Shots(shots)
```
Comment (albi3ro) [inline pennylane/gradients/finite_difference.py:340]:
```suggestion
```

File: pennylane/interfaces/execution.py [hunk 23]
```diff
@@ -481,8 +569,17 @@ def cost_fn(params, x):
...
             params.extend(tape.get_parameters(trainable_only=False))
         interface = qml.math.get_interface(*params)
+    if interface == "jax":
+        try:
+            from .jax import get_jax_interface_name
+        except ImportError as e:  # pragma: no-cover
+            raise qml.QuantumFunctionError(  # pragma: no-cover
+                "jax not found. Please install the latest "
+                "version of jax to enable the 'jax' interface."
+            ) from e
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:579]:
```suggestion
                "jax not found. Please install the latest " # pragma: no-cover
                "version of jax to enable the 'jax' interface." # pragma: no-cover
            ) from e # pragma: no-cover
```
